### PR TITLE
Don't capture wildcards if in closure or by-name

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes https://github.com/scala/bug/issues/9419

### CODE REVIEW

**Comment by dwijnand** [comment]
@odersky looking for any advice on perhaps a better place to do this - it's a shame to have to do it in two places.  Also it's doing it post-adapt, though that isn't unheard of.  Also, wasn't sure what to put in the error message 😄

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 1]
```diff
@@ -492,7 +493,20 @@ object ProtoTypes {
       val targ = cacheTypedArg(arg,
         typer.typedUnadapted(_, wideFormal, locked)(using argCtx),
         force = true)
-      typer.adapt(targ, wideFormal, locked)
+      val targ1 = typer.adapt(targ, wideFormal, locked)
+      if wideFormal eq formal then targ1
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:497]: Consider factoring this out into a separate function like `checkNoWildcardCaptureForCBN`

File: compiler/src/dotty/tools/dotc/typer/Typer.scala [hunk 2]
```diff
@@ -1568,6 +1568,10 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
                 else if ((tree.tpt `eq` untpd.ContextualEmptyTree) && mt.paramNames.isEmpty)
                   // Note implicitness of function in target type since there are no method parameters that indicate it.
                   TypeTree(defn.FunctionOf(Nil, mt.resType, isContextual = true, isErased = false))
+                else if mt.resType.match { case tp: AppliedType => tp.hasCaptureConversionArg case _ => false } then
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Typer.scala:1571]: That's a bit dense. Should we use an extractor (both here and in ProtoTypes)?

**Review by odersky — APPROVED** [review]
LGTM

**Pull request merged**